### PR TITLE
allow state changes from initialize

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
https://github.com/JuliaDiffEq/DifferentialEquations.jl/issues/194

Mostly mirroring https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/213

I'm not sure this is totally correct though since it may need to change some `integrator.integrator` values?

### CODE REVIEW

**Comment by devmotion** [comment]
I'm not sure... Shouldn't we do something similar to `handle_callback_modifiers!` and `reeval_internals_due_to_modification!` in the beginning of file https://github.com/JuliaDiffEq/DelayDiffEq.jl/blob/master/src/callbacks.jl? At least it seems reasonable to add a discontinuity of order 0 at the initial point (which might have been of order 1 before).

Moreover, we should probably replace most of the `integrator` with `dde_int` - `integrator` is just the ODE integrator that was used to construct `dde_int` and is responsible for interpolation.

**Comment by ChrisRackauckas** [comment]
We don't want to `add_steps!` though. Maybe we should add a trigger in there to avoid `add_steps!` specifically via an optional arg.

File: src/solve.jl [hunk 1]
```diff
@@ -296,8 +296,35 @@ function init(prob::AbstractDDEProblem{uType,tType,lType,isinplace}, alg::algTyp
 
     # set up additional initial values of newly created DDE integrator
     # (such as fsalfirst) and its callbacks
-    initialize!(dde_int)
-    initialize!(integrator.opts.callback, integrator.t, u, dde_int)
+
+    if initialize_integrator
```
Comment (devmotion) [inline src/solve.jl]: According to the failed tests it has to be added to list of keywords, as in OrdinaryDiffEq.

File: src/solve.jl [hunk 2]
```diff
@@ -296,8 +296,35 @@ function init(prob::AbstractDDEProblem{uType,tType,lType,isinplace}, alg::algTyp
...
+      integrator.u_modified = true
```
Comment (devmotion) [inline src/solve.jl]: I think we should (mostly) only operate on `dde_int` here and in the following lines. `integrator` is the wrong integrator, callbacks are initialized with `dde_int`.

**Comment by ChrisRackauckas** [comment]
Should be good now. @devmotion could you take a look?

**Comment by devmotion** [comment]
LGTM. With the recently tagged DiffEqBase version tests now also pass on Travis.

**Comment by ChrisRackauckas** [comment]
oh I see what happened to the tests. The fact that Travis Mac takes forever actually worked in our favor there lol.

**Pull request merged**